<!DOCTYPE html><html><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1"><style>body {
  max-width: 980px;
  border: 1px solid #ddd;
  outline: 1300px solid #fff;
  margin: 16px auto;
}

body .markdown-body
{
  padding: 45px;
}

@font-face {
  font-family: fontawesome-mini;
  src: url(data:font/woff;charset=utf-8;base64,d09GRgABAAAAABE0AA8AAAAAHWwAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAABHU1VCAAABWAAAADsAAABUIIslek9TLzIAAAGUAAAAQwAAAFY3d1HZY21hcAAAAdgAAACqAAACOvWLi0FjdnQgAAAChAAAABMAAAAgBtX/BGZwZ20AAAKYAAAFkAAAC3CKkZBZZ2FzcAAACCgAAAAIAAAACAAAABBnbHlmAAAIMAAABdQAAAjkYT9TNWhlYWQAAA4EAAAAMwAAADYQ6WvNaGhlYQAADjgAAAAfAAAAJAc6A1pobXR4AAAOWAAAACAAAAA0Kmz/7mxvY2EAAA54AAAAHAAAABwQPBJubWF4cAAADpQAAAAgAAAAIAEHC/NuYW1lAAAOtAAAAYQAAALxhQT4h3Bvc3QAABA4AAAAfgAAAMS3SYh9cHJlcAAAELgAAAB6AAAAhuVBK7x4nGNgZGBg4GIwYLBjYHJx8wlh4MtJLMljkGJgYYAAkDwymzEnMz2RgQPGA8qxgGkOIGaDiAIAJjsFSAB4nGNgZHZmnMDAysDAVMW0h4GBoQdCMz5gMGRkAooysDIzYAUBaa4pDA4Pwz+yMwf9z2KIYg5imAYUZgTJAQDcoQvQAHic7ZHNDYJAFIRnBXf94cDRIiyCKkCpwFCPJ092RcKNDoYKcN4+EmMPvpdvk539zQyAPYBCXEUJhBcCrJ5SQ9YLnLJe4qF5rdb+uWPDngNHTkta101pNyWa8lMhn6xx2dqUnW4q9YOIhAOOeueMSgsR/6ry+P7O5s6xVNg4chBsHUuFnWNJ8uZYwrw7chrsHXkODo7cB0dHOYCTY8kv0VE2WJKD6gOlWjsxAAB4nGNgQAMSEMgc9D8LhAESbAPdAHicrVZpd9NGFB15SZyELCULLWphxMRpsEYmbMGACUGyYyBdnK2VoIsUO+m+8Ynf4F/zZNpz6Dd+Wu8bLySQtOdwmpOjd+fN1czbZRJaktgL65GUmy/F1NYmjew8CemGTctRfCg7eyFlisnfBVEQrZbatx2HREQiULWusEQQ+x5ZmmR86FFGy7akV03KLT3pLlvjQb1V334aOsqxO6GkZjN0aD2yJVUYVaJIpj1S0qZlqPorSSu8v8LMV81QwohOImm8GcbQSN4bZ7TKaDW24yiKbLLcKFIkmuFBFHmU1RLn5IoJDMoHzZDyyqcR5cP8iKzYo5xWsEu20/y+L3mndzk/sV9vUbbkQB/Ijuzg7HQlX4RbW2HctJPtKFQRdtd3QmzZ7FT/Zo/ymkYDtysyvdCMYKl8hRArP6HM/iFZLZxP+ZJHo1qykRNB62VO7Es+gdbjiClxzRhZ0N3RCRHU/ZIzDPaYPh788d4plgsTAngcy3pHJZwIEylhczRJ2jByYCVliyqp9a6YOOV1WsRbwn7t2tGXzmjjUHdiPFsPHVs5UcnxaFKnmUyd2knNoykNopR0JnjMrwMoP6JJXm1jNYmVR9M4ZsaERCICLdxLU0EsO7GkKQTNoxm9uRumuXYtWqTJA/Xco/f05la4udNT2g70s0Z/VqdiOtgL0+lp5C/xadrlIkXp+ukZfkziQdYCMpEtNsOUgwdv/Q7Sy9eWHIXXBtju7fMrqH3WRPCkAfsb0B5P1SkJTIWYVYhWQGKta1mWydWsFqnI1HdDmla+rNMEinIcF8e+jHH9XzMzlpgSvt+J07MjLj1z7UsI0xx8m3U9mtepxXIBcWZ5TqdZlu/rNMfyA53mWZ7X6QhLW6ejLD/UaYHlRzodY3lBC5p038GQizDkAg6QMISlA0NYXoIhLBUMYbkIQ1gWYQjLJRjC8mMYwnIZhrC8rGXV1FNJ49qZWAZsQmBijh65zEXlaiq5VEK7aFRqQ54SbpVUFM+qf2WgXjzyhjmwFkiXyJpfMc6Vj0bl+NYVLW8aO1fAsepvH472OfFS1ouFPwX/1dZUJb1izcOTq/Abhp5sJ6o2qXh0TZfPVT26/l9UVFgL9BtIhVgoyrJscGcihI86nYZqoJVDzGzMPLTrdcuan8P9NzFCFlD9+DcUGgvcg05ZSVnt4KzV19uy3DuDcjgTLEkxN/P6VvgiI7PSfpFZyp6PfB5wBYxKZdhqA60VvNknMQ+Z3iTPBHFbUTZI2tjOBIkNHPOAefOdBCZh6qoN5E7hhg34BWFuwXknXKJ6oyyH7kXs8yik/Fun4kT2qGiMwLPZG2Gv70LKb3EMJDT5pX4MVBWhqRg1FdA0Um6oBl/G2bptQsYO9CMqdsOyrOLDxxb3lZJtGYR8pIjVo6Of1l6iTqrcfmYUl++dvgXBIDUxf3vfdHGQyrtayTJHbQNTtxqVU9eaQ+NVh+rmUfW94+wTOWuabronHnpf06rbwcVcLLD2bQ7SUiYX1PVhhQ2iy8WlUOplNEnvuAcYFhjQ71CKjf+r+th8nitVhdFxJN9O1LfR52AM/A/Yf0f1A9D3Y+hyDS7P95oTn2704WyZrqIX66foNzBrrblZugbc0HQD4iFHrY64yg18pwZxeqS5HOkh4GPdFeIBwCaAxeAT3bWM5lMAo/mMOT7A58xh0GQOgy3mMNhmzhrADnMY7DKHwR5zGHzBnHWAL5nDIGQOg4g5DJ4wJwB4yhwGXzGHwdfMYfANc+4DfMscBjFzGCTMYbCv6dYwzC1e0F2gtkFVoANTT1jcw+JQU2XI/o4Xhv29Qcz+wSCm/qjp9pD6Ey8M9WeDmPqLQUz9VdOdIfU3Xhjq7wYx9Q+DmPpMvxjLZQa/jHyXCgeUXWw+5++J9w/bxUC5AAEAAf//AA94nIVVX2hbZRQ/5/t7893s5ja9f7ouzdZ0TTqz3bRJmogbWya6bG6Cq0VbSV2ddIJjFtfIQHEig80Hda8yUN/0YQz8AyriiyD+xQd92R4HCnaCb3samnpumrpsCsLlfPf7zvedc37nL3CAtc/5W/wQZGA3tOBSY/g+TMjHmwzEoM1Q8+ZjRZY4oJhmBw5/YB6Za0yC5AkhlwA1A1yCBIBOwCII0Cj0U8BAMdUCzq05sKwkP7SlUY6fcJk4Fb/RyE79/6P5hjM/F4aZiXBoeMgzcqQ4Xi1hPqfDLG5FT+lchCVU3lYMyvuwhl1mqndQL0RsuloLywHtthLXI06OblTrhfWVnpSJ5+mwu/JdbtuN3IAnkW0LLMcRwaC7ktrlzridM6kVdyf9uO1UNBByI7JhwtG2sEwab07ORBeilWhqavJCqV0qzZTOl/7ZXQ5TbTcdcFelyGhhRDAQpdqp1FEX3w3cFTc1k9pJQkmm4ySCbSikxRP2QOfN+0tHS5MrpQuTU1Mk5nw0E5Xa0WvrOwDyGax9yB9ma6DAg82wHc43SAGTI4GjBWebOePAERFE8/AHaQpZASSTy8A4WwZiLQMQ82mFKATO0ILicRAoDm9p5P99E5b/fXG+kQYY3TYUuqmERWYoT0u/GNYL2q/4WB3LaVS+VynXsVYIcWw6DkCh3nX1D+VzlYN4LClF5yexSQos8exqZ3KVP+wtrC54u4Nznq6cq+xpMpUUnZ8FUYzE86ud0g28NOIv3Gj5/rmA3ABs7S/ywzFuQ4qyd6QxfNtiQIaEgp3w/entQg4Vcbqa16M5FfpeUB8t1+qeg7mI7cUyOe79wOk86gSxkVec4KPTX69++5x68Yubn5/F+w52z7u08sJX7fZXv8ekT/d2mILJxq6sn+SC6qEJknzLJCxyZEKwWVqYmAPBxBE/9DLeZiWHu7lcr/VytrCRuHojncNuTt9h46tmacmYisnSamdN2bZptcsmSysdVsy1PrOvOzF3xN64Rb937t/og9KHxYdcjIUqFAmIAHGHNzlns+RTPgeUYAQm9DwpNxfxbhhBHPaw3/gfTcXO2L+eJVIx5nsyGkvm9X4/f+bGkH45G0PaSjcMXTjcZyTvi3UdHoCDjQd3IDUVsgwYmUoJK/gp4JJxeRI0MKHZIkgynyIBqBTOUs6rOVCojvjZ4mCQz49ZMlMcp8QoYk6NoBfsxnJtsBohpa8iGJS+ZH7gU7NxME6cmF+t7cO9vB8d3jTWSct0ycW9ranXmolNDwmVkNnxe+8JtoztwS5rKJ0xWS95tQ/1zMYzg69MzUZnNtl1ofNbsml/OJm6f9wjRjpnu2o4MzHzn77IQkRd+1DjwMQ2pqSjGMMhyjrgTbBAKksuUm0iU7hI0aN2wOKOq7WYBSH0HGihj/jkiPxAfmwsEbfYrjMG+j3ij932Db/LV7I/xruNrhnroxjR9HRMb2nTvO0ZXOoHPk8H2ZhDPx93qcE/53sH5np/dkIP7zzhTVKdR/BAY/9ElkkR+A6lJGsqpJ4oQcTxpvBT3Kn58VkaJjgHyPEIws57xkaHh9KuVpDEpJZeMbZ5w/zBHi5NMQ4r5VphsFqID7TyB9eR4pX216c3AHxpdAwoqU9qg0ZJ6yVLKmMSz1iG2z27ifx18NkY0LPx1W/wCc2l5LrznrIsiKsqbmB78A9wIGx4tI8rjihVHJyY9pgMirenVq0yWg7Iw7eogG7ZgYM3qR9959A/fZkg6MnD/exlkmc+jWV4SB15XUR+eqC6l6ZmgPtN9z5JMfik05OV8ljylunJ4J+wA/FUaQSSKotsYsCWqaPBidBLcxkWx7XKFRIb45TGaEhjlF9uUVPqXOtcIwsXbBvfoZXIyRYFdkfnqjExH98xpnPczqzjX/uNdO1Y17Wpi5+6Ts8BXtjVFasp9KZ1mOiNbH65c5w6HgmyF2jFCZywM8mWjRc7T5Pmt0lRy7Y71+jYbpGyvwG4sH0XeJxjYGRgYADiwBB/53h+m68M3MwvgCIM1z5N/g6j///9v5H5BbMnkMvBwAQSBQCIcA9gAHicY2BkYGAO+p8FJF/8//v/F/MLBqAICuAFALYQB5kAeJxjfsHAwLwAiCNB+P9fbJjJmoGBMRUo/wKCAfO2EnQAAAAAANoBXgGcAgICVALaA1IDvAPkBAYEPARyAAEAAAANAF0ABAAAAAAAAgAUACQAcwAAAG4LcAAAAAB4nHWRzWrCQBSFT+pPqUIXLXTTzayKUohGKIibCoLuhbrrYtTRxCYZmYyKyz5Fd32HvlDfoO/QkziIFJtw9bvnnpl7ZwLgBt/wcHieGAf2UGd24Atcou+4RH3kuEweO66QXx1XyaHjGh6ROa7jFp/cwStfMVvhy7GHO+/e8QWuvcBxifqz4zL5xXGF/Oa4Sn53XMPE+3Bcx4P3M9DrvYmWoRWNQVN02kFXTPdCU4pSGQu5saE2meiLhU6timPtz3SSs9ypTCdqrJabWJoT5QQnymSRTkXgt0/UkUqVkVbN807ZdtmxdiEWRidi6HqItdErNbN+aO2612qd9sYAGmvsYRBhyUu0EGhQbfK/gzYCdElTOgSdB1eEFBIxFYkNV4RFJWPeZyyYpVQVHTHZx4y/yVGX2LGWFZri51TccUOn5B7nPefVCSPvGhVVwUl9znveO2KkhV8Wk82PZ8qwZf8OVcu1+fSmWCMw/HMOwXvKaysqM+p+cVuWag8tvv+c+xdd+4+teJxtjUEOwiAURJla24KliQfhUA2g/Sl+CKXx+loNrpzVezOLEY34Ron/0WhwQoszOvQYIKFwwQiNSbSBeO2SZ0tBP4j3zVjKNng32ZmtD1VVXCuOiw/pJ8S3WOU6l+K5UOTaDC4+2TjKMtN9KQf1ezLx/Sg/00FCvABHhjDjAAB4nGPw3sFwIihiIyNjX+QGxp0cDBwMyQUbGVidNjEwMmiBGJu5mBg5ICw+BjCLzWkX0wGgNCeQze60i8EBwmZmcNmowtgRGLHBoSNiI3OKy0Y1EG8XRwMDI4tDR3JIBEhJJBBs5mFi5NHawfi/dQNL70YmBhcADHYj9AAA) format('woff');
}

.markdown-body {
  font-family: sans-serif;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  color: #333333;
  overflow: hidden;
  font-family: "Helvetica Neue", Helvetica, "Segoe UI", Arial, freesans, sans-serif;
  font-size: 16px;
  line-height: 1.6;
  word-wrap: break-word;
}

.markdown-body a {
  background: transparent;
}

.markdown-body a:active,
.markdown-body a:hover {
  outline: 0;
}

.markdown-body b,
.markdown-body strong {
  font-weight: bold;
}

.markdown-body mark {
  background: #ff0;
  color: #000;
  font-style: italic;
  font-weight: bold;
}

.markdown-body sub,
.markdown-body sup {
  font-size: 75%;
  line-height: 0;
  position: relative;
  vertical-align: baseline;
}
.markdown-body sup {
  top: -0.5em;
}
.markdown-body sub {
  bottom: -0.25em;
}

.markdown-body h1 {
  font-size: 2em;
  margin: 0.67em 0;
}

.markdown-body img {
  border: 0;
}

.markdown-body hr {
  -moz-box-sizing: content-box;
  box-sizing: content-box;
  height: 0;
}

.markdown-body pre {
  overflow: auto;
}

.markdown-body code,
.markdown-body kbd,
.markdown-body pre,
.markdown-body samp {
  font-family: monospace, monospace;
  font-size: 1em;
}

.markdown-body input {
  color: inherit;
  font: inherit;
  margin: 0;
}

.markdown-body html input[disabled] {
  cursor: default;
}

.markdown-body input {
  line-height: normal;
}

.markdown-body input[type="checkbox"] {
  box-sizing: border-box;
  padding: 0;
}

.markdown-body table {
  border-collapse: collapse;
  border-spacing: 0;
}

.markdown-body td,
.markdown-body th {
  padding: 0;
}

.markdown-body .codehilitetable {
  border: 0;
  border-spacing: 0;
}

.markdown-body .codehilitetable tr {
  border: 0;
}

.markdown-body .codehilitetable pre,
.markdown-body .codehilitetable div.codehilite {
  margin: 0;
}

.markdown-body .linenos,
.markdown-body .code,
.markdown-body .codehilitetable td {
  border: 0;
  padding: 0;
}

.markdown-body td:not(.linenos) .linenodiv {
  padding: 0 !important;
}

.markdown-body .code {
  width: 100%;
}

.markdown-body .linenos div pre,
.markdown-body .linenodiv pre,
.markdown-body .linenodiv {
  border: 0;
  -webkit-border-radius: 0;
  -moz-border-radius: 0;
  border-radius: 0;
  -webkit-border-top-left-radius: 3px;
  -webkit-border-bottom-left-radius: 3px;
  -moz-border-radius-topleft: 3px;
  -moz-border-radius-bottomleft: 3px;
  border-top-left-radius: 3px;
  border-bottom-left-radius: 3px;
}

.markdown-body .code div pre,
.markdown-body .code div {
  border: 0;
  -webkit-border-radius: 0;
  -moz-border-radius: 0;
  border-radius: 0;
  -webkit-border-top-right-radius: 3px;
  -webkit-border-bottom-right-radius: 3px;
  -moz-border-radius-topright: 3px;
  -moz-border-radius-bottomright: 3px;
  border-top-right-radius: 3px;
  border-bottom-right-radius: 3px;
}

.markdown-body * {
  -moz-box-sizing: border-box;
  box-sizing: border-box;
}

.markdown-body input {
  font: 13px Helvetica, arial, freesans, clean, sans-serif, "Segoe UI Emoji", "Segoe UI Symbol";
  line-height: 1.4;
}

.markdown-body a {
  color: #4183c4;
  text-decoration: none;
}

.markdown-body a:hover,
.markdown-body a:focus,
.markdown-body a:active {
  text-decoration: underline;
}

.markdown-body hr {
  height: 0;
  margin: 15px 0;
  overflow: hidden;
  background: transparent;
  border: 0;
  border-bottom: 1px solid #ddd;
}

.markdown-body hr:before,
.markdown-body hr:after {
  display: table;
  content: " ";
}

.markdown-body hr:after {
  clear: both;
}

.markdown-body h1,
.markdown-body h2,
.markdown-body h3,
.markdown-body h4,
.markdown-body h5,
.markdown-body h6 {
  margin-top: 15px;
  margin-bottom: 15px;
  line-height: 1.1;
}

.markdown-body h1 {
  font-size: 30px;
}

.markdown-body h2 {
  font-size: 21px;
}

.markdown-body h3 {
  font-size: 16px;
}

.markdown-body h4 {
  font-size: 14px;
}

.markdown-body h5 {
  font-size: 12px;
}

.markdown-body h6 {
  font-size: 11px;
}

.markdown-body blockquote {
  margin: 0;
}

.markdown-body ul,
.markdown-body ol {
  padding: 0;
  margin-top: 0;
  margin-bottom: 0;
}

.markdown-body ol ol,
.markdown-body ul ol {
  list-style-type: lower-roman;
}

.markdown-body ul ul ol,
.markdown-body ul ol ol,
.markdown-body ol ul ol,
.markdown-body ol ol ol {
  list-style-type: lower-alpha;
}

.markdown-body dd {
  margin-left: 0;
}

.markdown-body code,
.markdown-body pre,
.markdown-body samp {
  font-family: Consolas, "Liberation Mono", Menlo, Courier, monospace;
  font-size: 12px;
}

.markdown-body pre {
  margin-top: 0;
  margin-bottom: 0;
}

.markdown-body kbd {
  background-color: #e7e7e7;
  background-image: -moz-linear-gradient(#fefefe, #e7e7e7);
  background-image: -webkit-linear-gradient(#fefefe, #e7e7e7);
  background-image: linear-gradient(#fefefe, #e7e7e7);
  background-repeat: repeat-x;
  border-radius: 2px;
  border: 1px solid #cfcfcf;
  color: #000;
  padding: 3px 5px;
  line-height: 10px;
  font: 11px Consolas, "Liberation Mono", Menlo, Courier, monospace;
  display: inline-block;
}

.markdown-body>*:first-child {
  margin-top: 0 !important;
}

.markdown-body>*:last-child {
  margin-bottom: 0 !important;
}

.markdown-body .headerlink {
  font: normal 400 16px fontawesome-mini;
  vertical-align: middle;
  margin-left: -16px;
  float: left;
  display: inline-block;
  text-decoration: none;
  opacity: 0;
  color: #333;
}

.markdown-body .headerlink:focus {
  outline: none;
}

.markdown-body h1 .headerlink {
  margin-top: 0.8rem;
}

.markdown-body h2 .headerlink,
.markdown-body h3 .headerlink {
  margin-top: 0.6rem;
}

.markdown-body h4 .headerlink {
  margin-top: 0.2rem;
}

.markdown-body h5 .headerlink,
.markdown-body h6 .headerlink {
  margin-top: 0;
}

.markdown-body .headerlink:hover,
.markdown-body h1:hover .headerlink,
.markdown-body h2:hover .headerlink,
.markdown-body h3:hover .headerlink,
.markdown-body h4:hover .headerlink,
.markdown-body h5:hover .headerlink,
.markdown-body h6:hover .headerlink {
  opacity: 1;
  text-decoration: none;
}

.markdown-body h1 {
  padding-bottom: 0.3em;
  font-size: 2.25em;
  line-height: 1.2;
  border-bottom: 1px solid #eee;
}

.markdown-body h2 {
  padding-bottom: 0.3em;
  font-size: 1.75em;
  line-height: 1.225;
  border-bottom: 1px solid #eee;
}

.markdown-body h3 {
  font-size: 1.5em;
  line-height: 1.43;
}

.markdown-body h4 {
  font-size: 1.25em;
}

.markdown-body h5 {
  font-size: 1em;
}

.markdown-body h6 {
  font-size: 1em;
  color: #777;
}

.markdown-body p,
.markdown-body blockquote,
.markdown-body ul,
.markdown-body ol,
.markdown-body dl,
.markdown-body table,
.markdown-body pre,
.markdown-body .admonition {
  margin-top: 0;
  margin-bottom: 16px;
}

.markdown-body hr {
  height: 4px;
  padding: 0;
  margin: 16px 0;
  background-color: #e7e7e7;
  border: 0 none;
}

.markdown-body ul,
.markdown-body ol {
  padding-left: 2em;
}

.markdown-body ul ul,
.markdown-body ul ol,
.markdown-body ol ol,
.markdown-body ol ul {
  margin-top: 0;
  margin-bottom: 0;
}

.markdown-body li>p {
  margin-top: 16px;
}

.markdown-body dl {
  padding: 0;
}

.markdown-body dl dt {
  padding: 0;
  margin-top: 16px;
  font-size: 1em;
  font-style: italic;
  font-weight: bold;
}

.markdown-body dl dd {
  padding: 0 16px;
  margin-bottom: 16px;
}

.markdown-body blockquote {
  padding: 0 15px;
  color: #777;
  border-left: 4px solid #ddd;
}

.markdown-body blockquote>:first-child {
  margin-top: 0;
}

.markdown-body blockquote>:last-child {
  margin-bottom: 0;
}

.markdown-body table {
  display: block;
  width: 100%;
  overflow: auto;
  word-break: normal;
  word-break: keep-all;
}

.markdown-body table th {
  font-weight: bold;
}

.markdown-body table th,
.markdown-body table td {
  padding: 6px 13px;
  border: 1px solid #ddd;
}

.markdown-body table tr {
  background-color: #fff;
  border-top: 1px solid #ccc;
}

.markdown-body table tr:nth-child(2n) {
  background-color: #f8f8f8;
}

.markdown-body img {
  max-width: 100%;
  -moz-box-sizing: border-box;
  box-sizing: border-box;
}

.markdown-body code,
.markdown-body samp {
  padding: 0;
  padding-top: 0.2em;
  padding-bottom: 0.2em;
  margin: 0;
  font-size: 85%;
  background-color: rgba(0,0,0,0.04);
  border-radius: 3px;
}

.markdown-body code:before,
.markdown-body code:after {
  letter-spacing: -0.2em;
  content: "\00a0";
}

.markdown-body pre>code {
  padding: 0;
  margin: 0;
  font-size: 100%;
  word-break: normal;
  white-space: pre;
  background: transparent;
  border: 0;
}

.markdown-body .codehilite {
  margin-bottom: 16px;
}

.markdown-body .codehilite pre,
.markdown-body pre {
  padding: 16px;
  overflow: auto;
  font-size: 85%;
  line-height: 1.45;
  background-color: #f7f7f7;
  border-radius: 3px;
}

.markdown-body .codehilite pre {
  margin-bottom: 0;
  word-break: normal;
}

.markdown-body pre {
  word-wrap: normal;
}

.markdown-body pre code {
  display: inline;
  max-width: initial;
  padding: 0;
  margin: 0;
  overflow: initial;
  line-height: inherit;
  word-wrap: normal;
  background-color: transparent;
  border: 0;
}

.markdown-body pre code:before,
.markdown-body pre code:after {
  content: normal;
}

/* Admonition */
.markdown-body .admonition {
  -webkit-border-radius: 3px;
  -moz-border-radius: 3px;
  position: relative;
  border-radius: 3px;
  border: 1px solid #e0e0e0;
  border-left: 6px solid #333;
  padding: 10px 10px 10px 30px;
}

.markdown-body .admonition table {
  color: #333;
}

.markdown-body .admonition p {
  padding: 0;
}

.markdown-body .admonition-title {
  font-weight: bold;
  margin: 0;
}

.markdown-body .admonition>.admonition-title {
  color: #333;
}

.markdown-body .attention>.admonition-title {
  color: #a6d796;
}

.markdown-body .caution>.admonition-title {
  color: #d7a796;
}

.markdown-body .hint>.admonition-title {
  color: #96c6d7;
}

.markdown-body .danger>.admonition-title {
  color: #c25f77;
}

.markdown-body .question>.admonition-title {
  color: #96a6d7;
}

.markdown-body .note>.admonition-title {
  color: #d7c896;
}

.markdown-body .admonition:before,
.markdown-body .attention:before,
.markdown-body .caution:before,
.markdown-body .hint:before,
.markdown-body .danger:before,
.markdown-body .question:before,
.markdown-body .note:before {
  font: normal normal 16px fontawesome-mini;
  -moz-osx-font-smoothing: grayscale;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  line-height: 1.5;
  color: #333;
  position: absolute;
  left: 0;
  top: 0;
  padding-top: 10px;
  padding-left: 10px;
}

.markdown-body .admonition:before {
  content: "\f056\00a0";
  color: 333;
}

.markdown-body .attention:before {
  content: "\f058\00a0";
  color: #a6d796;
}

.markdown-body .caution:before {
  content: "\f06a\00a0";
  color: #d7a796;
}

.markdown-body .hint:before {
  content: "\f05a\00a0";
  color: #96c6d7;
}

.markdown-body .danger:before {
  content: "\f057\00a0";
  color: #c25f77;
}

.markdown-body .question:before {
  content: "\f059\00a0";
  color: #96a6d7;
}

.markdown-body .note:before {
  content: "\f040\00a0";
  color: #d7c896;
}

.markdown-body .admonition::after {
  content: normal;
}

.markdown-body .attention {
  border-left: 6px solid #a6d796;
}

.markdown-body .caution {
  border-left: 6px solid #d7a796;
}

.markdown-body .hint {
  border-left: 6px solid #96c6d7;
}

.markdown-body .danger {
  border-left: 6px solid #c25f77;
}

.markdown-body .question {
  border-left: 6px solid #96a6d7;
}

.markdown-body .note {
  border-left: 6px solid #d7c896;
}

.markdown-body .admonition>*:first-child {
  margin-top: 0 !important;
}

.markdown-body .admonition>*:last-child {
  margin-bottom: 0 !important;
}

/* progress bar*/
.markdown-body .progress {
  display: block;
  width: 300px;
  margin: 10px 0;
  height: 24px;
  -webkit-border-radius: 3px;
  -moz-border-radius: 3px;
  border-radius: 3px;
  background-color: #ededed;
  position: relative;
  box-shadow: inset -1px 1px 3px rgba(0, 0, 0, .1);
}

.markdown-body .progress-label {
  position: absolute;
  text-align: center;
  font-weight: bold;
  width: 100%; margin: 0;
  line-height: 24px;
  color: #333;
  text-shadow: 1px 1px 0 #fefefe, -1px -1px 0 #fefefe, -1px 1px 0 #fefefe, 1px -1px 0 #fefefe, 0 1px 0 #fefefe, 0 -1px 0 #fefefe, 1px 0 0 #fefefe, -1px 0 0 #fefefe, 1px 1px 2px #000;
  -webkit-font-smoothing: antialiased !important;
  white-space: nowrap;
  overflow: hidden;
}

.markdown-body .progress-bar {
  height: 24px;
  float: left;
  -webkit-border-radius: 3px;
  -moz-border-radius: 3px;
  border-radius: 3px;
  background-color: #96c6d7;
  box-shadow: inset 0 1px 0 rgba(255, 255, 255, .5), inset 0 -1px 0 rgba(0, 0, 0, .1);
  background-size: 30px 30px;
  background-image: -webkit-linear-gradient(
    135deg, rgba(255, 255, 255, .4) 27%,
    transparent 27%,
    transparent 52%, rgba(255, 255, 255, .4) 52%,
    rgba(255, 255, 255, .4) 77%,
    transparent 77%, transparent
  );
  background-image: -moz-linear-gradient(
    135deg,
    rgba(255, 255, 255, .4) 27%, transparent 27%,
    transparent 52%, rgba(255, 255, 255, .4) 52%,
    rgba(255, 255, 255, .4) 77%, transparent 77%,
    transparent
  );
  background-image: -ms-linear-gradient(
    135deg,
    rgba(255, 255, 255, .4) 27%, transparent 27%,
    transparent 52%, rgba(255, 255, 255, .4) 52%,
    rgba(255, 255, 255, .4) 77%, transparent 77%,
    transparent
  );
  background-image: -o-linear-gradient(
    135deg,
    rgba(255, 255, 255, .4) 27%, transparent 27%,
    transparent 52%, rgba(255, 255, 255, .4) 52%,
    rgba(255, 255, 255, .4) 77%, transparent 77%,
    transparent
  );
  background-image: linear-gradient(
    135deg,
    rgba(255, 255, 255, .4) 27%, transparent 27%,
    transparent 52%, rgba(255, 255, 255, .4) 52%,
    rgba(255, 255, 255, .4) 77%, transparent 77%,
    transparent
  );
}

.markdown-body .progress-100plus .progress-bar {
  background-color: #a6d796;
}

.markdown-body .progress-80plus .progress-bar {
  background-color: #c6d796;
}

.markdown-body .progress-60plus .progress-bar {
  background-color: #d7c896;
}

.markdown-body .progress-40plus .progress-bar {
  background-color: #d7a796;
}

.markdown-body .progress-20plus .progress-bar {
  background-color: #d796a6;
}

.markdown-body .progress-0plus .progress-bar {
  background-color: #c25f77;
}

.markdown-body .candystripe-animate .progress-bar{
  -webkit-animation: animate-stripes 3s linear infinite;
  -moz-animation: animate-stripes 3s linear infinite;
  animation: animate-stripes 3s linear infinite;
}

@-webkit-keyframes animate-stripes {
  0% {
    background-position: 0 0;
  }

  100% {
    background-position: 60px 0;
  }
}

@-moz-keyframes animate-stripes {
  0% {
    background-position: 0 0;
  }

  100% {
    background-position: 60px 0;
  }
}

@keyframes animate-stripes {
  0% {
    background-position: 0 0;
  }

  100% {
    background-position: 60px 0;
  }
}

.markdown-body .gloss .progress-bar {
  box-shadow:
    inset 0 4px 12px rgba(255, 255, 255, .7),
    inset 0 -12px 0 rgba(0, 0, 0, .05);
}

/* MultiMarkdown Critic Blocks */
.markdown-body .critic_mark {
  background: #ff0;
}

.markdown-body .critic_delete {
  color: #c82829;
  text-decoration: line-through;
}

.markdown-body .critic_insert {
  color: #718c00 ;
  text-decoration: underline;
}

.markdown-body .critic_comment {
  color: #8e908c;
  font-style: italic;
}

.markdown-body .headeranchor {
  font: normal normal 16px fontawesome-mini;
  line-height: 1;
  display: inline-block;
  text-decoration: none;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.headeranchor:before {
  content: '\e157';
}

.markdown-body .task-list-item {
  list-style-type: none;
}

.markdown-body .task-list-item+.task-list-item {
  margin-top: 3px;
}

.markdown-body .task-list-item input {
  margin: 0 4px 0.25em -20px;
  vertical-align: middle;
}

/* Media */
@media only screen and (min-width: 480px) {
  .markdown-body {
    font-size:14px;
  }
}

@media only screen and (min-width: 768px) {
  .markdown-body {
    font-size:16px;
  }
}

@media print {
  .markdown-body * {
    background: transparent !important;
    color: black !important;
    filter:none !important;
    -ms-filter: none !important;
  }

  .markdown-body {
    font-size:12pt;
    max-width:100%;
    outline:none;
    border: 0;
  }

  .markdown-body a,
  .markdown-body a:visited {
    text-decoration: underline;
  }

  .markdown-body .headeranchor-link {
    display: none;
  }

  .markdown-body a[href]:after {
    content: " (" attr(href) ")";
  }

  .markdown-body abbr[title]:after {
    content: " (" attr(title) ")";
  }

  .markdown-body .ir a:after,
  .markdown-body a[href^="javascript:"]:after,
  .markdown-body a[href^="#"]:after {
    content: "";
  }

  .markdown-body pre {
    white-space: pre;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  .markdown-body pre,
  .markdown-body blockquote {
    border: 1px solid #999;
    padding-right: 1em;
    page-break-inside: avoid;
  }

  .markdown-body .progress,
  .markdown-body .progress-bar {
    -moz-box-shadow: none;
    -webkit-box-shadow: none;
    box-shadow: none;
  }

  .markdown-body .progress {
    border: 1px solid #ddd;
  }

  .markdown-body .progress-bar {
    height: 22px;
    border-right: 1px solid #ddd;
  }

  .markdown-body tr,
  .markdown-body img {
    page-break-inside: avoid;
  }

  .markdown-body img {
    max-width: 100% !important;
  }

  .markdown-body p,
  .markdown-body h2,
  .markdown-body h3 {
    orphans: 3;
    widows: 3;
  }

  .markdown-body h2,
  .markdown-body h3 {
    page-break-after: avoid;
  }
}
</style><style>/*GitHub*/
.codehilite {background-color:#fff;color:#333333;}
.codehilite .hll {background-color:#ffffcc;}
.codehilite .c{color:#999988;font-style:italic}
.codehilite .err{color:#a61717;background-color:#e3d2d2}
.codehilite .k{font-weight:bold}
.codehilite .o{font-weight:bold}
.codehilite .cm{color:#999988;font-style:italic}
.codehilite .cp{color:#999999;font-weight:bold}
.codehilite .c1{color:#999988;font-style:italic}
.codehilite .cs{color:#999999;font-weight:bold;font-style:italic}
.codehilite .gd{color:#000000;background-color:#ffdddd}
.codehilite .ge{font-style:italic}
.codehilite .gr{color:#aa0000}
.codehilite .gh{color:#999999}
.codehilite .gi{color:#000000;background-color:#ddffdd}
.codehilite .go{color:#888888}
.codehilite .gp{color:#555555}
.codehilite .gs{font-weight:bold}
.codehilite .gu{color:#800080;font-weight:bold}
.codehilite .gt{color:#aa0000}
.codehilite .kc{font-weight:bold}
.codehilite .kd{font-weight:bold}
.codehilite .kn{font-weight:bold}
.codehilite .kp{font-weight:bold}
.codehilite .kr{font-weight:bold}
.codehilite .kt{color:#445588;font-weight:bold}
.codehilite .m{color:#009999}
.codehilite .s{color:#dd1144}
.codehilite .n{color:#333333}
.codehilite .na{color:teal}
.codehilite .nb{color:#0086b3}
.codehilite .nc{color:#445588;font-weight:bold}
.codehilite .no{color:teal}
.codehilite .ni{color:purple}
.codehilite .ne{color:#990000;font-weight:bold}
.codehilite .nf{color:#990000;font-weight:bold}
.codehilite .nn{color:#555555}
.codehilite .nt{color:navy}
.codehilite .nv{color:teal}
.codehilite .ow{font-weight:bold}
.codehilite .w{color:#bbbbbb}
.codehilite .mf{color:#009999}
.codehilite .mh{color:#009999}
.codehilite .mi{color:#009999}
.codehilite .mo{color:#009999}
.codehilite .sb{color:#dd1144}
.codehilite .sc{color:#dd1144}
.codehilite .sd{color:#dd1144}
.codehilite .s2{color:#dd1144}
.codehilite .se{color:#dd1144}
.codehilite .sh{color:#dd1144}
.codehilite .si{color:#dd1144}
.codehilite .sx{color:#dd1144}
.codehilite .sr{color:#009926}
.codehilite .s1{color:#dd1144}
.codehilite .ss{color:#990073}
.codehilite .bp{color:#999999}
.codehilite .vc{color:teal}
.codehilite .vg{color:teal}
.codehilite .vi{color:teal}
.codehilite .il{color:#009999}
.codehilite .gc{color:#999;background-color:#EAF2F5}
</style><title>README</title></head><body><article class="markdown-body"><h1 id="sting">STing<a class="headerlink" href="#sting" title="Permanent link"></a></h1>
<h4 id="table-of-contents">Table of contents:<a class="headerlink" href="#table-of-contents" title="Permanent link"></a></h4>
<!-- MarkdownTOC levels="1,2,3" -->

<ul>
<li><a href="#requirements">Requirements</a></li>
<li><a href="#installation">Installation</a></li>
<li><a href="#quickstart-guide">Quickstart guide</a><ul>
<li><a href="#typing">Typing</a></li>
</ul>
</li>
<li><a href="#sting-applications">STing applications</a></li>
<li><a href="#sting-companion-scripts">STing companion scripts</a><ul>
<li><a href="#db_utilpy"><strong><code>db_util.py</code></strong></a></li>
<li><a href="#plot_kmer_depthr"><strong><code>plot_kmer_depth.R</code></strong></a></li>
</ul>
</li>
<li><a href="#using-sting">Using STing</a><ul>
<li><a href="#typing-mlst">Typing (MLST)</a></li>
<li><a href="#gene-detection">Gene Detection</a></li>
</ul>
</li>
<li><a href="#application-extended-help">Application extended help</a><ul>
<li><a href="#indexer">indexer</a></li>
<li><a href="#typer">typer</a></li>
</ul>
</li>
<li><a href="#seqan-version-211"><strong>SeqAn version:</strong> 2.1.1</a><ul>
<li><a href="#detector">detector</a></li>
</ul>
</li>
</ul>
<!-- /MarkdownTOC -->

<h2 id="requirements">Requirements<a class="headerlink" href="#requirements" title="Permanent link"></a></h2>
<ul>
<li>Linux OS</li>
<li>gcc &gt;= 4.8 and gcc &lt;= 5.3 </br>
    You may use gcc &gt;= 5.3 to compile STing but you will see multiple warning messages. Those warning messages will not interfere with the binary generation.</li>
<li>autotols (Ubuntu: autotools)<ul>
<li>Autoconf &gt;= 2.69</li>
<li>Automake &gt;= 1.15.1</li>
<li>Libtool &gt;= 2.4.6</li>
</ul>
</li>
<li>autotools developer headers (Ubuntu: autotools-dev)</li>
<li>zlib &gt;= 1.2.8 (Ubuntu: zlib)</li>
<li>zlib developer headers  (Ubuntu: zlib1g-dev)</li>
</ul>
<h2 id="installation">Installation<a class="headerlink" href="#installation" title="Permanent link"></a></h2>
<div class="codehilite"><pre>./autogen.sh
./configure
make
make install
</pre></div>

<p>By default, <code>make install</code> will install all the files in <code>/usr/local/bin</code>, <code>/usr/local/lib</code> etc.  You can specify an installation prefix other than <code>/usr/local</code> using the <code>--prefix</code> options from <code>./configure</code>, for instance <code>./configure --prefix=$HOME</code>.  Please check all the available options of <code>./configure</code> by executing <code>./configure --help</code>.</p>
<h2 id="quickstart-guide">Quickstart guide<a class="headerlink" href="#quickstart-guide" title="Permanent link"></a></h2>
<h3 id="typing">Typing<a class="headerlink" href="#typing" title="Permanent link"></a></h3>
<p>Preparing directory:</p>
<div class="codehilite"><pre>mkdir STing_demo
cd STing_demo
</pre></div>

<p>Downloading the <em>Neisseria spp.</em> MLST database from PubMLST.org and build a STing index from it:</p>
<div class="codehilite"><pre>db_util.py fetch --query &quot;Neisseria spp.&quot; --out_dir my_dbs --build_index
</pre></div>

<p>Downloading a WGS sample to analyze:</p>
<div class="codehilite"><pre>wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR026/ERR026529/ERR026529_1.fastq.gz ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR026/ERR026529/ERR026529_2.fastq.gz
</pre></div>

<p>Run STing typer:</p>
<div class="codehilite"><pre>typer -x my_dbs/neisseria_spp/db/index -1 ERR026529_1.fastq.gz -2 ERR026529_2.fastq.gz -s ERR026529 -c -a -d -t ERR026529.depth.tsv -y -o ERR026529.results.tsv --sensitive
</pre></div>

<h2 id="sting-applications">STing applications<a class="headerlink" href="#sting-applications" title="Permanent link"></a></h2>
<p>STing has three applications:</p>
<ul>
<li>
<dl>
<dt><strong><code>indexer</code></strong></dt>
<dd>Creates the indexes (databases) required to execute a locus-based typing analysis and detect genes.</dd>
</dl>
</li>
<li>
<dl>
<dt><strong><code>typer</code></strong></dt>
<dd>Predicts STs of a read sample based on an index built from a species-specific locus-based typing scheme (allelic profile table and sequences of observed alleles).</dd>
</dl>
</li>
<li>
<dl>
<dt><strong><code>detector</code></strong></dt>
<dd>Detects genes in a read sample based on an index built from sequences of a set of genes of interest.</dd>
</dl>
</li>
</ul>
<p>To explore the usage and available options for each tool, run the corresponding application using the option <code>-h</code> or <code>--help</code>:</p>
<div class="codehilite"><pre>indexer -h
typer -h
detector -h
</pre></div>

<h2 id="sting-companion-scripts">STing companion scripts<a class="headerlink" href="#sting-companion-scripts" title="Permanent link"></a></h2>
<p>STing includes two companion scripts:</p>
<h3 id="db_utilpy"><strong><code>db_util.py</code></strong><a class="headerlink" href="#db_utilpy" title="Permanent link"></a></h3>
<p>This script provides a set of utilities to download databases from PubMLST and build STing indices from them.</p>
<h3 id="plot_kmer_depthr"><strong><code>plot_kmer_depth.R</code></strong><a class="headerlink" href="#plot_kmer_depthr" title="Permanent link"></a></h3>
<p>This script Generates <strong>k</strong>-mer depth plots from output depth files generated by the <code>typer</code> and <code>detector</code> applications.</p>
<h4 id="requirements_1">Requirements:<a class="headerlink" href="#requirements_1" title="Permanent link"></a></h4>
<p><code>plot_kmer_depth.R</code> requires the following R packages:</p>
<ul>
<li>argparser</li>
<li>ggsci</li>
<li>gridExtra</li>
<li>RColorBrewer</li>
<li>stringr</li>
<li>svglite</li>
<li>tidyverse</li>
</ul>
<p>By default, <code>plot_kmer_depth.R</code> will try to install automatically the required packages to the personal R library directory (usually something like <code>~/R/x86_64-pc-linux-gnu-library/3.4</code>).  </p>
<h2 id="using-sting">Using STing<a class="headerlink" href="#using-sting" title="Permanent link"></a></h2>
<p>Next subsections present a brief description of the MLST analysis and gene detection workflows of STing. For the detailed help and usage of each application, please check the last section of this document (<a href="#extended-help-of-the-applications">Application extended help</a>).
In order to test the different STing tools, the folder <code>test</code> is included at the STing root directory. It contains the example files described below (configuration, profiles, and sequences) in addition to some read samples.</p>
<h3 id="typing-mlst">Typing (MLST)<a class="headerlink" href="#typing-mlst" title="Permanent link"></a></h3>
<p>Two main steps are required to execute an MLST analysis with STing: <em>Database indexing</em> and <em>ST prediction</em></p>
<h4 id="database-indexing">Database indexing<a class="headerlink" href="#database-indexing" title="Permanent link"></a></h4>
<ol>
<li>
<p>Create a config file that contains the path to loci and allelic profiles files. The format details for each file are the following:</p>
<h5 id="config-file">Config file<a class="headerlink" href="#config-file" title="Permanent link"></a></h5>
<p>A tab separated file that follows this format:</p>
<div class="codehilite"><pre>[loci]
locus1   relative/path/to/locusFile1.fa
locus2   relative/path/to/locusFile2.fa
[profile]
profile  relative/path/to/profileFile.txt
</pre></div>

<p>Blank lines and comments (lines starting with <code>#</code>) in this file, will be ignored. Paths are relative to the config file itself.
This is an example of a configuration file for the species <em>Neisseria spp.</em> (<code>test/mlst/pubmlst_db_files/Neisseria_spp/config.txt</code>):</p>
<div class="codehilite"><pre>[loci]
abcZ  abcZ.fa
adk adk.fa
aroE  aroE.fa
fumC  fumC.fa
gdh gdh.fa
pdhC  pdhC.fa
pgm pgm.fa
[profile]
Neisseria_spp profile.txt
</pre></div>

<h5 id="allele-sequence-file">Allele sequence file<a class="headerlink" href="#allele-sequence-file" title="Permanent link"></a></h5>
<p>A standard multi-FASTA file in which the id of each sequence consists of a locus name and an allele number separated by <code>_</code>:</p>
<div class="codehilite"><pre>&gt;abcZ_1
TTTGATACTGTTGCCGA...
&gt;abcZ_2
TTTGATACTGTTGCCGA...
</pre></div>

<h5 id="profile-file">Profile file<a class="headerlink" href="#profile-file" title="Permanent link"></a></h5>
<p>A tab separated file which contains each ST and its corresponding allelic profile (<code>test/mlst/pubmlst_db_files/Neisseria_spp/profile.txt</code>):</p>
<div class="codehilite"><pre>ST  abcZ  adk  aroE  fumC  gdh  pdhC  pgm  clonal_complex
1   1     3    1     1     1    1     3    ST-1 complex/subgroup I/II
2   1     3    4     7     1    1     3    ST-1 complex/subgroup I/II
3   1     3    1     1     1    23    13   ST-1 complex/subgroup I/II
4   1     3    3     1     4    2     3    ST-4 complex/subgroup IV
...
</pre></div>

<p>ST and loci columns (columns 1 to 7) are required in this file.</p>
</li>
<li>
<p>Build the database using the <code>indexer</code> tool:</p>
<div class="codehilite"><pre>indexer -m &lt;MODE&gt; -c &lt;CONFIG_FILE&gt; -p &lt;PREFIX_FILE&gt;
</pre></div>

<p>Example:</p>
<div class="codehilite"><pre>indexer -m MLST -c test/mlst/pubmlst_db_files/Neisseria_spp/config.txt -p test/mlst/dbs/Neisseria_spp/db
</pre></div>

<p>The command above will create an index for MLST analysis (mode MLST specified by <code>-m MLST</code>) using the MLST database files defined in the config file <code>test/mlst/pubmlst_db_files/Neisseria_spp/config.txt</code>. As a result, the indexer will generate a series of files named with the prefix <code>db</code> inside the directory <code>test/mlst/dbs/Neisseria_spp</code>.</p>
<p>The output of the command looks like this:</p>
<div class="codehilite"><pre>Loading sequences from sequences files:
N  Loci  #Seqs.  File
1  abcZ  1019    Neisseria_spp/abcZ.fa
2  adk   788     Neisseria_spp/adk.fa
3  aroE  1048    Neisseria_spp/aroE.fa
4  fumC  1125    Neisseria_spp/fumC.fa
5  gdh   1080    Neisseria_spp/gdh.fa
6  pdhC  1025    Neisseria_spp/pdhC.fa
7  pgm   1097    Neisseria_spp/pgm.fa
Total sequences loaded: 7182
Loading the profiles file... Done!
Creating and saving ESA index from loaded sequences... Done!
Index created successfully!
</pre></div>

</li>
</ol>
<h4 id="st-prediction">ST prediction<a class="headerlink" href="#st-prediction" title="Permanent link"></a></h4>
<p>Use the <code>typer</code>tool to predict the ST of a read set:</p>
<div class="codehilite"><pre>typer -x &lt;INDEX_PREFIX_FILENAME&gt; -1 &lt;FASTQ1&gt; -2 &lt;FASTQ2&gt; -k &lt;KMER_LENGTH&gt; -s &lt;SAMPLE_NAME&gt;
</pre></div>

<p>Example:</p>
<div class="codehilite"><pre>typer -x test/mlst/dbs/Neisseria_spp/db -1 test/mlst/samples/ERR017011_1.fastq.gz -2 test/mlst/samples/ERR017011_2.fastq.gz -s ERR017011
</pre></div>

<p>The command above will predict the ST of the read sample called <code>ERR017011</code> (<code>-s</code>), which comprises the input files <code>test/mlst/samples/ERR017011_1.fastq.gz</code> and <code>test/mlst/samples/ERR017011_2.fastq.gz</code> (specified by <code>-1</code> and <code>-2</code>), using the index located at the directory <code>test/mlst/dbs/Neisseria_spp</code> with the prefix <code>db</code> (<code>-x test/mlst/dbs/Neisseria_spp/db</code>).</p>
<p>The output of the previous command looks like this:</p>
<div class="codehilite"><pre>Sample  Line_type       Status  ST      abcZ    adk     aroE    fumC    gdh     pdhC    pgm     Total_k-mers  Total_reads      Input_files
ERR017011       allelic_profile st_predicted    5       1       1       2       1       3       2       3     76657    3597    ERR017011_1.fastq.gz,ERR017011_2.fastq.gz
</pre></div>

<p>By default, the typer application will send the header to <code>stderr</code>, and the prediction result to <code>stdout</code>. You can use the option <code>-o</code> to save the whole results (header and prediction) to a file, e.g., <code>-o my_results.tsv</code>.</p>
<h3 id="gene-detection">Gene Detection<a class="headerlink" href="#gene-detection" title="Permanent link"></a></h3>
<p>Two main steps are required to detect genes using STing: <em>Database building</em> and <em>Detection</em>.</p>
<h4 id="database-building">Database building<a class="headerlink" href="#database-building" title="Permanent link"></a></h4>
<ol>
<li>
<p>Create a config file that contains the path to gene files. The format details are the following:</p>
<h5 id="config-file_1">Config file<a class="headerlink" href="#config-file_1" title="Permanent link"></a></h5>
<p>A tab separated file with the following format:</p>
<div class="codehilite"><pre>[loci]
gene1   /path/to/geneFile1.fa
gene2   /path/to/geneFile2.fa
</pre></div>

<p>Blank lines and comments (lines starting with <code>#</code>) in this file, will be ignored.
Note that there are no <code>[profile]</code> section for a configuration file for gene detection. If the file contains this section, the <code>indexer</code> tool will show an error message.
This is an example of a configuration file of AMR genes (<code>test/amr/amr_db_files/set_01/config.txt</code>):</p>
<div class="codehilite"><pre>[loci]
erm erm.fasta
ksg ksg.fasta
pen pen.fasta
qac qac.fasta
aac2ic  aac2ic.fasta
aph6id  aph6id.fasta
bl2a_1  bl2a_1.fasta
ermb  ermb.fasta
mepa  mepa.fasta
pbp2b pbp2b.fasta
pbp2x pbp2x.fasta
tetpa tetpa.fasta
</pre></div>

<h5 id="gene-sequence-file">Gene sequence file<a class="headerlink" href="#gene-sequence-file" title="Permanent link"></a></h5>
<p>A standard multi-FASTA file in which the id is the name of the gene. In case of having genes with the same name, you should add a number to the name separated by <code>_</code>:</p>
<div class="codehilite"><pre>&gt;pen_1
TTTGATACTGTTGCCGA...
&gt;pen_2
TTTGATACTGTTGCCGA...
</pre></div>

</li>
<li>
<p>Build the database using the <code>indexer</code> tool:</p>
<div class="codehilite"><pre>indexer -m &lt;MODE&gt; -c &lt;CONFIG_FILE&gt; -p &lt;PREFIX_FILE&gt;
</pre></div>

<p>Example:</p>
<div class="codehilite"><pre>indexer -m GDETECT -p databases/amr -c test/amr/amr_db_files/set_01/config.txt
</pre></div>

<p>The command above will create an index for gene detection (mode GDETECT in the config file <code>test/amr/amr_db_files/set_01/config.txt</code>. As a result, the indexer will create a series of files named with the prefix <code>amr</code> inside the directory <code>databases</code>.</p>
<p>The output of the command looks like this:</p>
<div class="codehilite"><pre>Loading sequences from sequences files:
N     Loci    #Seqs.  File
1     aac2ic  1       set_01/aac2ic.fasta
2     aph6id  1       set_01/aph6id.fasta
3     bl2a_1  1       set_01/bl2a_1.fasta
4     erm     1       set_01/erm.fasta
5     ermb    1       set_01/ermb.fasta
6     ksg     2       set_01/ksg.fasta
7     mepa    1       set_01/mepa.fasta
8     pbp2b   1       set_01/pbp2b.fasta
9     pbp2x   1       set_01/pbp2x.fasta
10    pen     4       set_01/pen.fasta
11    qac     1       set_01/qac.fasta
12    tetpa   1       set_01/tetpa.fasta
Total loaded sequences: 16
Creating and saving ESA index from loaded sequences...
Index created successfuly!
</pre></div>

</li>
</ol>
<h4 id="detection">Detection<a class="headerlink" href="#detection" title="Permanent link"></a></h4>
<p>You must use the <code>detector</code> tool to detect genes in a read set:</p>
<div class="codehilite"><pre>detector -x &lt;INDEX_PREFIX_FILENAME&gt; -1 &lt;FASTQ1&gt; -2 &lt;FASTQ2&gt; -k &lt;KMER_LENGTH&gt;
</pre></div>

<p>Example:</p>
<div class="codehilite"><pre>detector -x databases/amr -1 test/amr/samples/GCF_000008805.fasta.40.1.fq.gz -2 test/amr/samples/GCF_000008805.fasta.40.2.fq.gz -k 30 -s GCF_000008805
</pre></div>

<p>The command above will detect presence/absence (1/0) of the genes from the database, in the read sample called <code>GCF_000008805</code> (<code>-s</code>) which comprises the input files <code>test/amr/samples/GCF_000008805.fasta.40.1.fq.gz</code> and <code>test/amr/samples/GCF_000008805.fasta.40.2.fq.gz</code> (specified by <code>-1</code> and <code>-2</code>), using the index located at the directory <code>databases</code> with the prefix <code>amr</code> (<code>-x databases/amr</code>). Additionally, the tool will use <em>k</em>-mers of size 30 (<code>-k 30</code>) to process the input reads.</p>
<p>The output of the previous command looks like this:</p>
<div class="codehilite"><pre>Sample  Line_type ermC  ksga1 ksga2 pbp2b pen1  pen2  pen3  pen4  qacE1 Total_hits  Total_kmers Total_reads Input_files 
GCF_000008805 presence  1 1 0 1 1 1 1 1 1 288106170395  1396  GCF_000008805.fasta.40.1.fq.gz,GCF_000008805.fasta.40.2.fq.gz
</pre></div>

<p>By default, the detector application will send the header to <code>stderr</code>, and the prediction result to <code>stdout</code>. </p>
<h2 id="application-extended-help">Application extended help<a class="headerlink" href="#application-extended-help" title="Permanent link"></a></h2>
<h3 id="indexer">indexer<a class="headerlink" href="#indexer" title="Permanent link"></a></h3>
<h4 id="synopsis">Synopsis<a class="headerlink" href="#synopsis" title="Permanent link"></a></h4>
<p><strong>indexer</strong> [OPTIONS] -c <CONFIG_FILE>  </p>
<h4 id="description">Description<a class="headerlink" href="#description" title="Permanent link"></a></h4>
<p>STing <strong>indexer</strong> creates indexes (DBs) required by the STing <strong>typer</strong> and <strong>detector</strong> programs for loci-based typing analysis and detecting genes, respectively, from NGS raw sequence reads.</p>
<dl>
  <dt>-h, --help</dt>
    <dd>Display the help message.</dd>
  <dt>--version</dt>
    <dd>Display version information.</dd>
  <dt>-c, --config-file CONFIG_FILE</dt>
    <dd>A tab delimited file whith names and paths to the typing scheme files (see the FILE FORMAT DETAILS section below).</dd>
  <dt>-p, --db-prefix PREFIX</dt>
    <dd>Filename prefix for the DB files to be created. You can specify a folder structure here to store your DB at a particular location, e.g., path/to/my/db/prefix. Default: name of the config file CONFIG_FILE</dd>
  <dt>-m, --mode MODE</dt>
    <dd>Indexing mode. Valid options: MLST, GDETECT. Select MLST to create a database for MLST analysis or GDETECT to create a database for gene detection. Default: _MLST_.</dd>
</dl>

<h4 id="file-format-details">File format details<a class="headerlink" href="#file-format-details" title="Permanent link"></a></h4>
<h5 id="config_file">CONFIG_FILE<a class="headerlink" href="#config_file" title="Permanent link"></a></h5>
<p>A tab separated file with the name and location of files for creating a DB.
Config files for MLST DBs (MLST mode) must have two sections: <strong>[loci]</strong> that describes names and paths to alleles sequence files for each locus, and <strong>[profile]</strong> that describes the name and path to the profile file.
Config files for gene detection DBs (GDETECT mode), only require the [loci] section.
An example of a config file for a MLST DB is as follows:</p>
<div class="codehilite"><pre>[loci]
locus1 relative/path/to/locusFile1
locus2 relative/path/to/locusFile2
locusN relative/path/to/locusFileN
[profile]
profile relative/path/to/profileFile
</pre></div>

<p>Paths are relative to the config file itself. Blank lines and comments (lines starting with &lsquo;#&rsquo;) will be ignored.</p>
<h5 id="allele-sequence-file_1">Allele sequence file<a class="headerlink" href="#allele-sequence-file_1" title="Permanent link"></a></h5>
<p>A standard multi-FASTA file (.fa or .fasta) in which each sequence description must be the locus name and the allele number separated by &lsquo;_&rsquo;:</p>
<div class="codehilite"><pre>&gt;abcZ_1
TTTGATACTGTTGCCGA...
&gt;abcZ_2
TTTGATACTGTTGCCGA...
</pre></div>

<h5 id="profile-file_1">Profile file<a class="headerlink" href="#profile-file_1" title="Permanent link"></a></h5>
<p>A tab separated file that contains the ST and the corresponding allelic profile:</p>
<div class="codehilite"><pre>ST abcZ adk aroE fumC gdh pdhC pgm clonal_complex
1 1 3 1 1 1 1 3 ST-1 complex/subgroup I/II
2 1 3 4 7 1 1 3 ST-1 complex/subgroup I/II
3 1 3 1 1 1 23 13 ST-1 complex/subgroup I/II
4 1 3 3 1 4 2 3 ST-4 complex/subgroup IV
</pre></div>

<h4 id="version">Version<a class="headerlink" href="#version" title="Permanent link"></a></h4>
<p><strong>Last update:</strong> 09/11/2018<br />
<strong>indexer version:</strong> 0.23.2<br />
<strong>SeqAn version:</strong> 2.1.1</p>
<hr />
<h3 id="typer">typer<a class="headerlink" href="#typer" title="Permanent link"></a></h3>
<h4 id="synopsis_1">Synopsis<a class="headerlink" href="#synopsis_1" title="Permanent link"></a></h4>
<p><strong>typer</strong> -x <INDEX_PREFIX_FILENAME> -1 <FASTQ1> [OPTIONS]  </p>
<h4 id="description_1">Description<a class="headerlink" href="#description_1" title="Permanent link"></a></h4>
<p>STing <strong>typer</strong> is an ultrafast assembly- and alignment-free program for sequence typing directly from whole-genome raw sequence reads. STing <strong>typer</strong> is based on k-mer frequencies, and works with locus-based typing schemes like those defined in the traditional MLST method and its derivatives (e.g, rMLST, cgMLST or wgMLST). STing <strong>typer</strong> requires an index (DB) created with the STing <strong>indexer</strong> program.</p>
<dl>
    <dt>-h, --help</dt>
    <dd>Display the help message.</dd>
    <dt>--version</dt>
    <dd>Display version information.</dd>
</dl>

<h5 id="required-input-parameters">Required input parameters:<a class="headerlink" href="#required-input-parameters" title="Permanent link"></a></h5>
<dl>
    <dt>-x, --index-prefix INDEX_PREFIX_FILENAME</dt>
    <dd>Index prefix filename.</dd>
    <dt>-1, --fastq-1-files FASTQ1</dt>
    <dd>Files with #1 mates, paired with files in _FASTQ2_. Valid file extensions are _.fq_, _.fastq_ (uncompressed fastq), and _.gz_ (gzip'ed fastq).</dd>
</dl>

<h5 id="input-options">Input options:<a class="headerlink" href="#input-options" title="Permanent link"></a></h5>
<dl>
    <dt>-2, --fastq-2-files FASTQ2</dt>
    <dd>Files with #2 mates, paired with files in _FASTQ1_. Valid file extensions are _.fq_, _.fastq_ (uncompressed fastq), and _.gz_ (gzip'ed fastq).</dd>
    <dt>-s, --sample-name SAMPLE_NAME</dt>
    <dd>Name of the sample to be analized.</dd>
    <dt>-k, --kmer-length KMER_LENGTH</dt>
    <dd>Length of the k-mers to process the input reads. Default: _30_.</dd>
    <dt>-n, --n-top-alleles N_TOP_ALLELES</dt>
    <dd>Number of top alleles by k-mer frequencies, from which best alleles will be called. Default: _2_.</dd>
</dl>

<h5 id="output-options">Output options:<a class="headerlink" href="#output-options" title="Permanent link"></a></h5>
<dl>
    <dt>-c, --kmer-counts</dt>
    <dd>Select and print the normalized k-mer counts at each locus (k-mer hits divided by allele length).</dd>
    <dt>-a, --allele-cov</dt>
    <dd>Select to calculate and print the percent of the length of each allele that is covered by k-mer hits. Note that this calculation will require more execution time.</dd>
    <dt>-d, --kmer-depth</dt>
    <dd>Select to calculate and print the average k-mer depth of each allele. Note that this calculation will require more execution time.</dd>
    <dt>-t, --k-depth-file KMER_DEPTH_FILENAME</dt>
    <dd>Output filename to save the detailed per-base k-mer depth data.</dd>
    <dt>-e, --ext-k-depth-file EXT_KMER_DEPTH_FILENAME</dt>
    <dd>Output filename to save the extended per-base k-mer depth data.</dd>
    <dt>-o, --output-file OUTPUT_FILENAME</dt>
    <dd>Output filename to save the typing results (instead of stdout).</dd>
    <dt>-y, --print-tidy</dt>
    <dd>Select to print results in a tidy format.</dd>
    <dt>-v, --verbose</dt>
    <dd>Select to print informative messages (to stderr).</dd>
</dl>

<h5 id="presets">Presets:<a class="headerlink" href="#presets" title="Permanent link"></a></h5>
<dl>
    <dt>--fast</dt>
    <dd>(Default). Select to set the _fast_ mode to call alleles based solely on k-mer frequencies. The best allele of each locus is that with the highest k-mer hit frequency.</dd>
    <dt>--sensitive</dt>
    <dd>Select to set the _sensitive_ mode to call alleles based on k-mer frequencies and coverage information. The best allele of each locus is that with the highest k-mer hit frequency and the highest allele coverage. Allele ties are solved selecting the allele with the minimum k-mer depth standard deviation.</dd>
</dl>

<h4 id="note-on-fastq1-and-fastq2">Note on FASTQ1 and FASTQ2<a class="headerlink" href="#note-on-fastq1-and-fastq2" title="Permanent link"></a></h4>
<p><em>FASTQ1</em> and <em>FASTQ2</em> can be comma-separated lists (no whitespace) and can be specified many times, e.g., -1 file1.fq,file2.fq -1 file3.fq</p>
<h4 id="output-conventions">Output conventions<a class="headerlink" href="#output-conventions" title="Permanent link"></a></h4>
<h5 id="status-values">Status values:<a class="headerlink" href="#status-values" title="Permanent link"></a></h5>
<dl>
    <dt>st_predicted</dt>
    <dd>ST inferred from k-mer hits and profiles table.</dd>
    <dt>no_kmer_hits</dt>
    <dd>There are no k-mer hits for one or more of the loci to infer a profile and its associated ST. Probable causes:</dd>
    <dd>1) k-mer length not adequate (too long),</dd>
    <dd>2) low quality data/too many N's in the data.</dd>
    <dt>no_st_in_table</dt>
    <dd>There is no ST associated to the inferred allelic profile. Probable causes:</dd>
    <dd>1) k-mer length not adequate (too short),</dd>
    <dd>2) this could be a new allelic profile.</dd>
</dl>

<h5 id="other-valuessymbols">Other values/symbols:<a class="headerlink" href="#other-valuessymbols" title="Permanent link"></a></h5>
<dl>
    <dt>NA</dt>
    <dd>No ST associated or no k-mer hits for any allele of a locus.</dd>
    <dt>*</dt>
    <dd>Low confidence. Allele predicted with a length coverage below 100%. Probable causes:</dd>
    <dd>1) No enough k-mer hits to cover the whole length of the allele</dd>
    <dd>2) No enough k-mer depth in some part along the range [10, length-10] bp of the allele to consider it as covered.</dd>
</dl>

<h4 id="version_1">Version<a class="headerlink" href="#version_1" title="Permanent link"></a></h4>
<p><strong>Last update:</strong> 02/28/2019<br />
<strong>typer version:</strong> 0.24.2<br />
<strong>SeqAn version:</strong> 2.1.1</p>
<hr />
<h3 id="detector">detector<a class="headerlink" href="#detector" title="Permanent link"></a></h3>
<h4 id="synopsis_2">Synopsis<a class="headerlink" href="#synopsis_2" title="Permanent link"></a></h4>
<p><strong>detector</strong> -x <INDEX_PREFIX_FILENAME> -1 <FASTQ1> [OPTIONS]  </p>
<h4 id="description_2">Description<a class="headerlink" href="#description_2" title="Permanent link"></a></h4>
<p>STing <strong>detector</strong> is an ultrafast assembly- and alignment-free program for detecting genes directly from NGS raw sequence reads. STing <strong>detector</strong> is based on k-mer frequencies. STing <strong>detector</strong> requires an index (DB) created with the STing <strong>indexer</strong> program (using the GDETECT mode).
<dl>
  <dt>-h, &ndash;help</dt>
    <dd>Display the help message.</dd>
  <dt>&ndash;version</dt>
    <dd>Display version information.</dd>
</dl></p>
<h4 id="required-input-parameters_1">Required input parameters:<a class="headerlink" href="#required-input-parameters_1" title="Permanent link"></a></h4>
<p><dl>
  <dt>-x, &ndash;index-prefix INDEX_PREFIX_FILENAME</dt>
    <dd>Database prefix filename.</dd>
  <dt>-1, &ndash;fastq-1-files FASTQ1</dt>
    <dd>Files with #1 mates, paired with files in <em>FASTQ1</em>.</dd>
</dl></p>
<h4 id="input-options_1">Input options:<a class="headerlink" href="#input-options_1" title="Permanent link"></a></h4>
<p><dl>
  <dt>-2, &ndash;fastq-2-files FASTQ2</dt>
    <dd>Files with #2 mates, paired with files in <em>FASTQ2</em>.</dd>
  <dt>-s, &ndash;sample-name SAMPLE_NAME</dt>
    <dd>Name of the sample to be analized.</dd>
  <dt>-k, &ndash;kmer-length KMER_LENGTH</dt>
    <dd>Length of the k-mers to process the input reads. Default: <em>30</em>.</dd>
  <dt>-r, &ndash;threshold THRESHOLD</dt>
    <dd>Minimum length coverage (%) required to consider a gene as present in a sample. In range [1.0..100.0]. Default: <em>75</em>.</dd>
</dl></p>
<h5 id="output-options_1">Output options:<a class="headerlink" href="#output-options_1" title="Permanent link"></a></h5>
<dl>
  <dt>-c, --kmer-counts</dt>
    <dd>Select to print the number of k-mer matches at each gene.</dd>
  <dt>-p, --kmer-perc</dt>
    <dd>Select to print the percentage of k-mer matches from the total of processed k-mers.</dd>
  <dt>-g, --gene-cov</dt>
    <dd>Select to print the percent of the gene length that is covered by the corresponding k-mer matches.</dd>
  <dt>-d, --kmer-depth</dt>
    <dd>Select to print the mean k-mer depth of each gene.</dd>
  <dt>-y, --print-tidy</dt>
    <dd>Select to print results in a tidy format.</dd>
  <dt>-t, --k-depth-file KMER_DEPTH_FILENAME</dt>
    <dd>Output filename to save the detailed per-base k-mer depth data.</dd>
  <dt>-v, --verbose</dt>
    <dd>Select to print informative messages (to stderr). Default non verbose.</dd>
</dl>

<p><em>FASTQ1</em> and <em>FASTQ2</em> can be comma-separated lists (no whitespace) and can be specified many times. E.g. -1 file1.fq,file2.fq -1 file3.fq</p>
<h4 id="version_2">Version<a class="headerlink" href="#version_2" title="Permanent link"></a></h4>
<p><strong>Last update:</strong> 10/13/2018<br />
<strong>detector version:</strong> 0.23.3<br />
<strong>SeqAn version:</strong> 2.1.1</p></article></body></html>